### ISSUE TITLE
org.freedesktop.gstreamer:gst1-java-core:0.9.4 obviously not able to overcome network boundaries

### ISSUE THREAD

**Issue by neilyoung** [centricular/gstwebrtc-demos#117]
...which in fact contradicts the purpose and power of WebRTC.

OK, what are the facts?

I'm having a Raspberry PI here, which is providing WebRTC via UV4L and a little proxy I wrote in order to connect to my Signalling Server. After some back and forth the last days (see https://github.com/centricular/gstwebrtc-demos/issues/113)  I was able to make it run with GST webrtcbin from within a Java app on the other end.

In this scenario the PI is issuing an OFFER on demand (this is the only way UV4L works), the OFFER is forwarded by the Signalling Server to the Java app using GStreamer webrtcbin, which ANSWERs. In between the ICE exchange happens.

This scenario works fine here on my network, where the PI is located. 

I was asking a friend in Paris to do the same from the same Java app. He could not connect. 

Compared to my traces I'm missing two lines in his trace, I usually see here on my machine:

```
0:00:11.283791000  2018 0x7faa032c8320 TRACE              webrtcbin gstwebrtcbin.c:722:_collate_ice_connection_states:<sendrecv> ICE connection state: 0x8
0:00:11.283796000  2018 0x7faa032c8320 TRACE              webrtcbin gstwebrtcbin.c:763:_collate_ice_connection_states:<sendrecv> returning connected
```

So for me it looks like as if no ICE candidate pair could reach "connected" in the "french connection" :) This is astonishing, because accessing the same PI from the same network in France using a browser as other party works perfectly (If anybody is interested to give this a try I could share some info about what to do in order to achieve that, it is simple. You just need a browser).

So facts:
- I can achieve full connectivity with the PI, either from a browser or from the Java app, but I'm on the same network with both clients.
- My friend in Paris can achieve connectivity using the browser.
- My friend in Paris can't achieve that using a Java app, which utilises GStreamer webrtcbin.

Today I tried to boil it down to the least expensive setup:

- I forced the PI to UDP candidates only
- I made the GST Java app able to ignore IPv6 TCP candidates and work with IPv4 UDP candidates only 
- I put the PI into a separate Wifi sub net, separated from the net the PC is using. There is usually no direct intra-net traffic possible ("Guest-Network" vs. "Home-Network")

As a matter of fact I was NOT ABLE to establish a connection from the Java app to the PI. In contrary I could easily establish a connection using a browser based web app on the same machine, the Java app was running. So it _is_ possible.  It goes without saying, that I can reach full connectivity with the Java app, if I put my PC into the same network, the PI uses.

I have created a log using `GST_DEBUG=3,*webrtc*:7;G_MESSAGES_DEBUG=libnice,libnice-stun`. It is a bit chatty, but it clearly shows, that GST is unable to establish a connection. Just the reason is still unknown. 
My verdict is: GStreamer webrtcbin is unable to traverse NAT/FW, which is uhmm... strange, to be careful.

Here is the log: My PC was having 192.168.188.24, the PI 192.168.189.30 (note the different class C networks). SDP and ICE exchange organized, no ICE connectivity in the end.

I'm having here a successful trace. Both on the same network. 

PI: 192.168.189.30
PC: 192.168.189.21

[gst_successful.log](https://github.com/centricular/gstwebrtc-demos/files/3362041/gst_successful.log)

And here we have the unsuccessful case. 

PI: 192.168.189.30
PC: 192.168.188.24

[gst_unsuccessful.log](https://github.com/centricular/gstwebrtc-demos/files/3362044/gst_unsuccessful.log)

**Comment by sdroege** [comment]
What's "GStreamer WebRTCBin 0.9.4"? There is no such GStreamer version. Which GStreamer (and gst-plugins-bad) version, and which libnice version are you using?

I assume direct sending of UDP packets between the two subnets does not work, so in that case you'll need to proxy via a TURN server. Are there any ICE candidates between the two networks that would work, e.g. via a TURN server? STUN can only work to poke holes into NATs, but I don't think you have a NAT between your two subnets? Does the browser application you mention use a TURN server?

You say you only keep the UDP ICE candidates, but if no UDP packets can be directly sent between the two subnets then this is not going to work obviously.

**Comment by neilyoung** [comment]
Well, this is what I'm using:

    compile "org.freedesktop.gstreamer:gst1-java-core:0.9.4"

Sorry, for me this is GStreamer WebRTCBin, but in fact it is "gst1-java-core".

> Does the browser application you mention use a TURN server?

No. But let me check, what the candidate pair finally is.

> You say you only keep the UDP ICE candidates, but if no UDP packets can be directly sent between the two subnets then this is not going to work obviously.

Basically this is currently my last line of defence. I came from a fully blown ICE "nest" on both sides: TCP/UDP/IPv4/IPv6 - it didn't work either. Can easily return to this.

**Comment by neilyoung** [comment]
![image](https://user-images.githubusercontent.com/731020/60714389-47db5280-9f1b-11e9-85dc-c9f187e94858.png)

The browser uses the public IP, what is totally correct. In fact, why is GST not doing this? The public IP is known from STUN.

**Comment by neilyoung** [comment]
GStreamer Core Library version 1.14.4

I don't currently have it on top of my head, how to determine the versions of gst-plugins-bad and libnice 

I'm no macOS, but my french friend is on Ubuntu, so this seems to be not the issue.

**Comment by sdroege** [comment]
> The browser uses the public IP, what is totally correct. In fact, why is GST not doing this? The public IP is known from STUN.

So if you filter out all candidates but the ones for the public IP, does it work then? I.e. is the problem only that for some reason this one is not tried, or does it even not work when having the public IP as only candidate left?

In either case, this needs debug logs from libnice and is likely a problem in there. So knowing which version you're using would also be useful.

I'd also recommend updating to GStreamer (and all the plugin modules) 1.16.0 as there were many changes in the webrtc stack.

**Comment by neilyoung** [comment]
As said: I can (and will) easily revert that filtering. It was just a test to minimize the complexity a bit. As you know, the GST core fires 50-80 candidates, it is pretty hard to keep the overview. I will right now make a log using ALL candidates.

> In either case, this needs debug logs from libnice and is likely a problem in there. So knowing which version you're using would also be useful.

No problem, if I just would know, how :)

OK, update possible, but let's do this as last attempt. I remember I had to update to 1.14.4 using source builds, otherwise significant parts of the plugins-bad have not been installed (e.g. libnice).

`
brew edit gst-plugins-bad
brew reinstall --build-from-source gst-plugins-bad
`

OK, next to come a trace with full ICE cands

**Comment by neilyoung** [comment]
As said: No success either with the whole shebang of candidates. 

[gst-allcandidates-no-success.log](https://github.com/centricular/gstwebrtc-demos/files/3362184/gst-allcandidates-no-success.log)

**Comment by sdroege** [comment]
> As said: I can (and will) easily revert that filtering. It was just a test to minimize the complexity a bit. As you know, the GST core fires 50-80 candidates, it is pretty hard to keep the overview. I will right now make a log using ALL candidates.

It would make sense, for testing, to limit the candidates to the ones with the public IP. As that's known to work with the browsers. That way we would know if the problem is the selection done in libnice, or if it fails to use an known-to-work candidate for some reason.

Nonetheless, of course if there were no bugs you ideally wouldn't do any filtering at all here as that way you'd allow e.g. a direct connection between peers in the same subnet while going via different route otherwise.



> OK, update possible, but let's do this as last attempt. I remember I had to update to 1.14.4 using source builds, otherwise significant parts of the plugins-bad have not been installed (e.g. libnice).
> 
> `brew edit gst-plugins-bad brew reinstall --build-from-source gst-plugins-bad`

https://gstreamer.freedesktop.org/data/pkg/osx/1.16.0/ has everything in the latest versions in a known-to-work configuration.

**Comment by neilyoung** [comment]
OK, I'll give it a shot. I suppose, I don't have to install the devel packages?

**Comment by sdroege** [comment]
You can get libnice debug output by setting `NICE_DEBUG=all` in the environment btw.

**Comment by sdroege** [comment]
> OK, I'll give it a shot. I suppose, I don't have to install the devel packages?

I don't know how the Java bindings work. Probably not. 

You might also have to do something so that they use the binaries installed from the package.

**Comment by sdroege** [comment]
Ah your log already contains libnice debug output. This seems to be the relevant line at the end:

```
(unknown:3682): libnice-DEBUG: 12:10:55.481: Agent 0x7f9acf2d1720 : Failed pair is [192.168.188.24]:64614 --> [80.133.185.223]:43259
```

You said that this is supposed to work but there's no response to the STUN request sent here after waiting 12 seconds. Might be worth checking with wireshark on both machines to check if it is actually sent out, if it actually arrives on the other side, and what the other side is actually doing there.

libnice logs from both sides would also help.

**Comment by neilyoung** [comment]
> You can get libnice debug output by setting NICE_DEBUG=all in the environment btw.

I already have libnice traces now

EDIT: Crosspost

**Comment by neilyoung** [comment]
OK, this needs a bit of a setup change here. The PI is headless, no Wireshark. I bet nobody wants to mess with tcpdump. But since it also not works, if I use a browser on one end and the Java app on the other, if both are on different networks, I could easily launch another PC in the other network and try to call it from here. That should give me the means to trace with WS on both ends.

Needs a while

**Comment by alexlapa** [comment]
@neilyoung ,
https://nice.freedesktop.org/libnice/NiceAgent.html#nice-agent-add-local-address
> the GST core fires 50-80 candidates

Pehaps, machine running gst app has a lot of local networks? If you know the correct interface that will be used, then you can pass it directly to libnice agent, although im not sure that webrtc-bin exposes agent used. 

Ive had such problems before (running libnice in docker with network-mode=host), so i had to use custom libnice: https://github.com/alexlapa/libnice , which [detects correct network interface](https://github.com/alexlapa/libnice/commit/6c5fca132c39b4c641dfa90b0208f7e3e7024f2a)

And [this](https://nice.freedesktop.org/libnice/NiceAgent.html#nice-agent-add-local-address) function will allow you to disable automatic interface discovery and specify correct one directly, if webrtc-bin provides direct access to NiceAgent.

**Comment by neilyoung** [comment]
I installed your package, but it didn't change the version to 1.16.

**Comment by sdroege** [comment]
> I installed your package, but it didn't change the version to 1.16.

As said above, you'll most likely need to do something so that the Java bindings make use of that version. I'm not using macOS or the Java bindings, so no idea what is needed there.

> OK, this needs a bit of a setup change here. The PI is headless, no Wireshark. I bet nobody wants to mess with tcpdump. But since it also not works, if I use a browser on one end and the Java app on the other, if both are on different networks, I could easily launch another PC in the other network and try to call it from here. That should give me the means to trace with WS on both ends.

Thanks, that would be useful. Wireshark capture on both sides, and GStreamer/libnice debug log of the same run from on side.

Alternatively tcpdump would also work of course, both write the same format.

**Comment by neilyoung** [comment]
I just installed it out of the package for all users, using my admin rights. Then I checked the version and it was still 1.14.4

gst-launch-1.0 --gst-version
GStreamer Core Library version 1.14.4

I probably need to get rid of the version installed by `brew`

**Comment by sdroege** [comment]
> I probably need to get rid of the version installed by `brew`

You at least need to add the other version to your `PATH` somehow. It's going to be installed in `/Libraries/Frameworks/GStreamer.framework` or similar, with `gst-launch-1.0` being inside the `bin` directory there.

Removing the brew version shouldn't be needed.

**Comment by sdroege** [comment]
But even then, I have no idea how to tell the Java bindings to use the version from that path on macOS. On other platforms that would also be easy by setting an environment variable, but on macOS this is more tricky as shared libraries are linked with the whole path... so I assume the Java bindings are going to continue looking for the libraries in `/usr/local/...` or wherever brew is installing them.

**Comment by neilyoung** [comment]
Seems so. I have this now at command line, but no change in behavior:

```
~ $ which gst-launch-1.0
/Library/Frameworks/GStreamer.framework/Commands/gst-launch-1.0
~ $ gst-launch-1.0 --gst-version
GStreamer Core Library version 1.16.0

```

**Comment by neilyoung** [comment]
Upgrading via brew seems to work

```
==> Upgrading 5 outdated packages:
gstreamer 1.14.4 -> 1.16.0_1, gst-plugins-base 1.14.4 -> 1.16.0_1, gst-plugins-good 1.14.4_1 -> 1.16.0_4, gst-plugins-bad 1.14.4 -> 1.16.0_1, gst-plugins-ugly 1.14.4_2 -> 1.16.0_1

```

**Comment by neilyoung** [comment]
Upgrade successful, patient dead :)

`0:00:09.825274000 33875 0x7f81a46b3f50 WARN     GST_ELEMENT_FACTORY gstelementfactory.c:456:GstElement *gst_element_factory_make(const gchar *, const gchar *): no such element factory "webrtcbin"!
`

**Comment by sdroege** [comment]
So apparently the new brew version didn't come with the webrtc plugin by default and you need to do something. Check the brew documentation

**Comment by neilyoung** [comment]
Yes, I installed plugins-bad manually and got the webrtcbin. But now videotestsrc is missing. I guess I try a fresh update

**Comment by sdroege** [comment]
Yeah brew does not seem to be very user-friendly in that regard, I saw such problems reported often in various places. That's why I suggested the official GStreamer macOS binaries, but unfortunately due to macOS being very developer-unfriendly that does not play well with the Java bindings. The general brokeness of macOS is something you'll have to figure out yourself, I'm afraid. After that we can look at the connection issues.

**Comment by neilyoung** [comment]
I think I will make it. Not the first time. But it needs time. Will return later. Thanks for your patience so far.

**Comment by neilyoung** [comment]
For me BTW macOS is the best amongst all. I also have Windows and Linux experiences and devices, was forced to switch to Mac 10 years ago with the first iPhone SDK and I would never ever make any other platform my main work horse. And I'm doing a lot of things: From programming mobile devices, web apps, drivers, IoT things --- with respect to stability and reliability there is no better OS in the market. I'm not a fan boy, but I would never give up my MacBook Pro  :) Just my opinion :) And by that I can really cope with some of those issues, which I always was able to fix so far. Not a big deal.

**Comment by neilyoung** [comment]
A bit disappointing now. 1.16.0 is installed and there is no change in the behaviour with respect to the issue here. Also this issue is still present too (https://github.com/centricular/gstwebrtc-demos/issues/115)

**Comment by sdroege** [comment]
Good, now we're at least on an up to date version and not looking at any of the many bugs that were fixed in the meantime :) Sad that this specific one was not solved by this, but let's continue debugging then.

Can you check which version of libnice you're using (and update it if needed). Then update your code to only pass through the candidates that are known to work (AFAIU the public IP?). And then get a debug log as before from the GStreamer side, and a wireshark packet capture from both sides?

**Comment by neilyoung** [comment]
OK, my setup now:

- A web app on the left in network 1 (192.168.189.x)
- The java app on the right in network 2 (192.168.188.x)
- STUN configured on both sides to be 

`urls: ["stun:54.152.156.46:3478", "stun:stun.l.google.com:19302"] 
`
- Java app OFFERs to web app. No video
- Having a full WS STUN trace for both sides. PCAP format OK?

Java app trace (including libnice traces):
[gst-allcandidates-fail.log](https://github.com/centricular/gstwebrtc-demos/files/3362623/gst-allcandidates-fail.log)

Wireshark traces (STUN Filter set) for both sides:
[wireshark-traces.zip](https://github.com/centricular/gstwebrtc-demos/files/3362627/wireshark-traces.zip)

**Comment by neilyoung** [comment]
Hmm. Realized that I should return to the old trace setup. This trace doesn't show STUN.

Disregard the post above. Will produce something new

**Comment by sdroege** [comment]
>     * PCAP format OK?

Yes

**Comment by sdroege** [comment]
> Hmm. Realized that I should return to the old trace setup. This trace doesn't show STUN.

It also does not include any libnice debug output

**Comment by neilyoung** [comment]
> It also does not include any libnice debug output

Yes, I was following your advice 

> You can get libnice debug output by setting NICE_DEBUG=all in the environment btw.

that doesn't give anything good. But wait a second please, I now have reduced the Java traces to my traces and the libnice/stun traces only. It looks clearer now.

One second. Assembling the files

**Comment by sdroege** [comment]
> Wireshark traces (STUN Filter set) for both sides:
> [wireshark-traces.zip](https://github.com/centricular/gstwebrtc-demos/files/3362627/wireshark-traces.zip)

The Java app sends STUN binding requests to 192.168.189.31 and 80.133.185.223 (i.e. your IPs) and not to the ones you configured ("stun:54.152.156.46:3478", "stun:stun.l.google.com:19302").

The Web app sends STUN binding requests to the given server and those are also succeeding. So the problem here seems to be that libnice tries to do things directly instead of poking the STUN servers you configured. How did you configure them in the Java application btw?

**Comment by neilyoung** [comment]
```
webRTCBin.setStunServer("stun:stun.l.google.com:19302");
webRTCBin.setStunServer("stun:54.152.156.46:3478");
```

Not much documentation found.

**Comment by sdroege** [comment]
> ```
> webRTCBin.setStunServer("stun:stun.l.google.com:19302");
> webRTCBin.setStunServer("stun:54.152.156.46:3478");
> ```
> 
> 
> Not much documentation found.

The C documentation says:
>   stun-server         : The STUN server of the form stun://hostname:port

So you probably use the wrong format here.

**Comment by neilyoung** [comment]
Nice. This was, what I was changing when I started this test. The initial form was this (commented now):

`//webRTCBin.setStunServer("stun://stun.l.google.com:19302");`

But even with this or even without any configuration - it does not work.

**Comment by neilyoung** [comment]
OK; having this now

 ```
                               webRTCBin.setStunServer("stun://stun.l.google.com:19302");
                                webRTCBin.setStunServer("stun://54.152.156.46:3478");

```

Same result. Do you want to have the traces?

**Comment by sdroege** [comment]
You can only set a single one btw, setting the second time will unset the first. There seems to be no way currently to configure multiple STUN servers.

> Same result. Do you want to have the traces?

Yes, now we're closer to things being correct at least.

**Comment by neilyoung** [comment]
OK, configured just this in Java

 `                               webRTCBin.setStunServer("stun://54.152.156.46:3478");
`

[gst-allcandidates-fail.log](https://github.com/centricular/gstwebrtc-demos/files/3362712/gst-allcandidates-fail.log)
[wireshark-traces.zip](https://github.com/centricular/gstwebrtc-demos/files/3362715/wireshark-traces.zip)

For me it seems, that the Java app - as you said - first tried to bind to a STUN server in my network and then uses the configured one. Maybe ways too late.

**Comment by neilyoung** [comment]
Yes, this seems to be the case: LIBNICE is giving up after a couple of attempts to unrelated and not configured servers. But _before_ it attempts to bind to the correct server:

Look at the time stamp:


```
(unknown:66723): libnice-DEBUG: 14:35:37.873: Agent 0x7fbf9aa40d20 : Retransmissions failed, giving up on pair 0x7fbf9aee6230
(unknown:66723): libnice-DEBUG: 14:35:37.873: Agent 0x7fbf9aa40d20 : Failed pair is [192.168.188.24]:54962 --> [192.168.189.31]:64288
(unknown:66723): libnice-DEBUG: 14:35:37.873: Agent 0x7fbf9aa40d20 : pair 0x7fbf9aee6230 state FAILED (candidate_check_pair_fail)


```

The first binding request to the proper STUN server goes out 14:35:54 (according to WS)

**Comment by neilyoung** [comment]
Something to bite :) I'm out for today. Many thanks for your patience and your competent help. Hopefully we will be able to find a solution or a work around.

**Comment by sdroege** [comment]
```
(unknown:66723): libnice-DEBUG: 14:37:10.065: agent_recv_message_unlocked: Packet passed fast STUN validation but failed slow validation.
```
is probably the problem here.

You could try filtering out all the other irrelevant candidates to make sure this is indeed not a timing related problem. But to me this looks like there's something wrong in libnice here. Should be reported here: https://gitlab.freedesktop.org/libnice/libnice/issues/new

**Comment by neilyoung** [comment]
Wouldn’t you follow my explanation? Libnice messes too long with the wrong servers. At least this is what I see in the WS trace

**Comment by sdroege** [comment]
> Wouldn’t you follow my explanation? Libnice messes too long with the wrong servers. At least this is what I see in the WS trace

Doesn't look like a problem on the GStreamer side, but what I see is that libnice is complaining that validation of the messages sent from the web side failed. And that's definitely a problem :)

If libnice needs to long and wastes time doing useless stuff that would also be a bug in libnice though.

**Comment by neilyoung** [comment]
Thanks. Three issues and all have a dead en

**Comment by neilyoung** [comment]
I don’t understand what is so special with the error message you were quoting. The error is given more than 30 seconds before Libnice  contacts the proper server. So what does that say?

**Comment by sdroege** [comment]
> Thanks. Three issues and all have a dead en

3? One is a bug in gstreamer-java, this one looks like a bug in libnice. So not really a dead-end but has to be followed up on at a different place.

What's the third?

> I don’t understand what is so special with the error message you were quoting. The error is given more than 30 seconds before Libnice contacts the proper server. So what does that say

That something goes wrong in libnice apparently. It seems to not understand the messages coming from the proper server either but seems to receive them, and then libnice/webrtcbin simply give up because no working candidate could be found. So whatever the actual problem, libnice fails to create a working connection here.

**Comment by sdroege** [comment]
Ah you mean https://github.com/centricular/gstwebrtc-demos/issues/113 as the third one? That's a bug in GStreamer, which I've asked you to forward but you didn't yet. It should error out properly here instead of exploding on old, unsupported SDP formats for datachannel.

**Comment by neilyoung** [comment]
Yeah, the follow up places concern me. Most likely there are not half as responsive persons as you are ;)

**Comment by sdroege** [comment]
Well, this repository here only contains some webrtc example applications using GStreamer. Highly likely that any problems you find are not in the code here but in GStreamer or any of the other dependencies used here.

**Comment by neilyoung** [comment]
I don’t think I will make some extra noise regarding 113. I can circumvent and the workaround even improves the experience. Also for the gstreamer-java thing. As long as if it would work with 0.9.4 - why bother? This last one is ugly and I will for sure have to follow up, since the situation now disqualifies the entire GST approach.

**Comment by sdroege** [comment]
> I don’t think I will make some extra noise regarding 113. I can circumvent and the workaround even improves the experience. Also for the gstreamer-java thing.

Well if you're not interested in getting things fixed properly I don't see any point in spending more time on helping you debug things any further.

> As long as if it would work with 0.9.4 - why bother?

Sure if you want to stay at an old version forever, go for it.

**Comment by neilyoung** [comment]
Well I thank you a lot, but I don’t think it is necessary to blame me for having other priorities. Since you know about the issues why don’t you place them in the appropriate lists?  These are all not my issues. For the data channel - what’s wrong with switching it off to circumvent problems at gstreamer side, even if I don’t need it?  And for the version issue - well, if ever comes the time and I will be forced to update to 1.1.0 maybe then ... But to go through all the hassle and extra work for a non productive POC? Honestly, my world improvement times are long time gone, I don’t have any ambitions anymore.  No need for this twist in our discussion. 

You know what? I even have the opportunity to forget about gst entirely, since I’m having a fully working solution based on a headless chromium usage. 

So anyway: no need for unnecessary bad vibes after a successful day. And many thanks again, really.

**Comment by sdroege** [comment]
That's perfectly fine if that's enough for you, but don't expect any free help from anybody in that case. Naive from me to assume you'd like to get things fixed properly, which was my motivation to help you debugging *your* problems in my spare time to begin with :)

**Comment by neilyoung** [comment]
So then:  your help deserves a personal favor. I will consider to contribute as far as i can. In my spare time :)

**Comment by neilyoung** [comment]
Filed everything.

**Comment by neilyoung** [comment]
@alexlapa That sounds interesting. I think I'm using a libnice version installed via brew:

```
~ $ brew info libnice
libnice: stable 0.1.16 (bottled)
GLib ICE implementation
https://wiki.freedesktop.org/nice/
/usr/local/Cellar/libnice/0.1.16_1 (66 files, 1.5MB) *
  Poured from bottle on 2019-07-05 at 13:26:04
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/libnice.rb
==> Dependencies
Build: pkg-config ✔
Required: glib ✔, gnutls ✔, gstreamer ✔
==> Analytics
install: 157 (30 days), 399 (90 days), 1,240 (365 days)
install_on_request: 101 (30 days), 298 (90 days), 989 (365 days)
build_error: 0 (30 days)
~ $ 

```

Your lib is a fork of Kurento's libnice, which also somehow is derived from the freedesktop thing, isn't it? I can't see, that webrtcbin does expose any direct access to the libnice functions

**Comment by neilyoung** [comment]
@sdroege: Plot twist: The issue is solved. And the solution completely exonerates LIBNICE.

The problems began, when I started to integrate a Raspberry PI, using the UV4L provided WebRTC solution. This UV4L solution is probably a bit outdated and has some "other" specialities. So it is unable to react on an incoming OFFER - it is only able to act as OFFERING party. 

Then I had two problems:

- It issues an outdated SDP format for data channels, which made GStreamer raise a critical warning (and stop working, BTW). I could circumvent by disabling the data channel at all, since I'm not using it

- The webrtcbin plugin delivers ICE candidates without the "sdpMid" property. This is a problem for UV4L, so in order to overcome this, I added "sdpMid=video" to the entire ICE message w/o thinking twice. That removed  the problem at UV4L side, but in fact it caused a REAL BIG problem for all other clients, because it MOST LIKELY makes them discarding received ICE candidates tagged with this "sdpMid" property "out of the blue".

This morning I did remember, that there was a time, when I was easily able to work across NAT boundaries using the GStreamer webrtcbin plugin. That time I was using the example Java application from here https://github.com/centricular/gstwebrtc-demos. I re-enabled it this morning and boom - it worked. And the only difference compared to my current Java app was: It did not contain all the changes I made in order to make UV4L run...

So everything works now and I'm convinced it will also work with the main stream LIBNICE distribution; even with 1.14.1.

Thanks again for your help
